src/strands/tools/decorator.py

```diff
@@ -444,6 +441,34 @@ async def stream(self, tool_use: ToolUse, invocation_state: dict[str, Any], **kw
                 "content": [{"text": f"Error: {error_type} - {error_msg}"}],
             }
 
+    async def _stream(self, validated_input: dict[str, Any], invocation_state: dict[str, Any]) -> Any:
+        """Execute the tool function based on type.
+
+        Args:
+            validated_input: Validated input parameters for the tool function.
+            invocation_state: Context for the tool invocation.
+
+        Returns:
+            The result of the tool function execution.
+        """
+        # "Too few arguments" expected, hence the type ignores
+
+        if inspect.iscoroutinefunction(self._tool_func):
+            return await self._tool_func(**validated_input)  # type: ignore
+
+        thread_pool = invocation_state.get("thread_pool")
```

Comment:
Is this something that would be placed in the new FunctionContext that @dbschmigelski is putting together?